Add `setup` command to make initial setup a single step

wmanley opened the pull request:
TODO:

- [x] Tidy up
- [x] Install keyring dependency and explicitly prompt for auth key
- [x] Update vscode config to enable pylint, test discovery and snapshot on save
- [x] Prompt for node-id
- [x] Check that git credentials are set up.  `GIT_TERMINAL_PROMPT=0 git ls-remote`?
- `.env` Tests
- `--vscode` tests
- [x] Test running setup with no requests installed.
- [x] Test `_update_env`

wmanley committed:
Add `./stbt_rig.py setup` to create a correctly configured venv for a test-pack

We re-exec after installing requests and keyring so we don't need them
installed for setup to run, but can still use them later in setup.

wmanley committed:
Import requests only when needed

This way `stbt_rig.py setup` won't require requests as it will be
responsible for installing requests in the first place.

wmanley committed:
setup: Prompt for auth token

This will be saved thanks to the Python keyring package.

wmanley committed:
setup: Prompt the user to select a default node-id to use

This is saved in the `.env` file that VSCode reads before executing
commands.  VSCode supports multiple environment files and makes it easy
to switch between them, so it makes it easy to switch between nodes.

wmanley committed:
setup: Add --vscode option to update `.vscode/settings.json`

...according to our recommendations.

wmanley committed:
setup: Pass in node_id rather than node

We don't want to construct a `Portal` and `Node` as a prerequsite of setup
as we don't know whether we even have requests setup yet.

wmanley committed:
setup: Ensure the user has name and email setup

So `snapshot` works.  In the future I'd like to make this automatic,
pulling the information from the portal, but for now that requires OAuth,
and I don't want to have to figure that out just for this feature.

wmanley committed:
stbt_rig.py setup: Fixes for Windows

venv's work a little differently on Windows, including having a different
filesystem layout.

We also need to worry about "Execution Aliases".  Execution aliases
are how Windows prints:

> Python was not found; run without arguments to install from the
> Microsoft Store, or disable this shortcut from Settings > Manage App
> Execution Aliases.

when you run `python` at the command line while Python isn't installed.

Unfortunately they break our python detection as
`distutils.spawn.find_executable` finds it and thinks that it's real
Python when it's just a stub:
https://blogs.windows.com/windowsdeveloper/2017/07/05/command-line-activation-universal-windows-apps/#QBVyMPUOpI9SMbFI.97

wmanley committed:
Tests: Revamp test_pack fixture

This will be useful for the setup tests I'm about to write.

wmanley committed:
Tests: Move fixture portal_mock into conftests.py

For reuse by setup tests.

wmanley committed:
Tests: Add tests for stbt_rig setup

wmanley committed:
Tests: Add __main__ so we can play with the test environment

drothlis commented on stbt_rig.py:
```diff
@@ -493,6 +502,58 @@ def cmd_snapshot(args, node):
+        # We re-exec ourselves into the venv now that we've installed all our
+        # dependencies
+        os.execv("%s/.venv/bin/python" % root,
+                 ["%s/.venv/bin/python" % root, this_stbt_rig] + sys.argv[1:])
```
These path separators work on Windows, do they?

drothlis commented on stbt_rig.py:
```diff
@@ -485,6 +505,236 @@ def cmd_snapshot(args, node):
+            "--tb=no", "--capture=no",
+            "tests"
+        ],
+        "runOnSave.commands": [
```
```suggestion
        # This requires the "Run on Save" extension by "pucelle":
        "runOnSave.commands": [
```

drothlis commented on stbt_rig.py:
```diff
@@ -485,6 +505,236 @@ def cmd_snapshot(args, node):
+                die("Can't find python %s in PATH" % python_version)
+
+            # Create venv
+            subprocess.check_call(python + ['-m', 'venv', '.venv'], cwd=root)
```
This only works with Python 3 because 2.7 doesn't have a `venv` module. Is the plan for this `setup` command to be Python3 only? I'm in favour of that but we'd need to make it explicit by checking the stbt.conf's python version and issuing a user-friendly error message.

wmanley replied:
Good spot. Yeah, ok I'll drop Python 2 support from setup and add an error message.  In the unlikely event that there's actually demand I can add it in again with virtualenv support.

drothlis commented on stbt_rig.py:
```diff
@@ -485,6 +505,236 @@ def cmd_snapshot(args, node):
+        else:
+            os.execv(cmd[0], cmd)
+    else:
+        # Install ourselves as a python module in the venv:
```
```suggestion
        # Install stbt_rig as a python module in the venv.
        #
        # We use stbt_rig as a pytest plugin to run tests from inside VS Code,
        # like this: `pytest -p stbt_rig`. However VS Code doesn't execute
        # pytest directly; it executes pytest via a wrapper script. In late
        # 2020 that wrapper script started removing the workspace root
        # directory (the test-pack checkout) from sys.path before invoking
        # pytest, so we need to install stbt_rig into the venv.
```

drothlis reviewed (commented):
I still don't like the idea of hiding the venv creation from the user, creating our own mechanism for installing dependencies and managing their versions, etc etc. pip & venv seem like a crucial part of the python ecosystem that our users just need to learn. Home-grown wrappers that "manage" a well-known mature tool for you, always offer a bad experience. But I know you're set on this so we can give it a go. You'll need to update the documentation at https://stb-tester.com/manual/ide-configuration

wmanley commented:
Thanks for the review.

> I still don't like the idea of hiding the venv creation from the user, creating our own mechanism for installing dependencies and managing their versions, etc etc. pip & venv seem like a crucial part of the python ecosystem that our users just need to learn.

Setting up your development environment is the first step to get started with our tool. We want the barrier to entry be as low as possible and have the greatest chance of succeeding without requiring a bunch of debugging. The alternative for our customers who aren't already experienced python package wranglers isn't learn pip, it's to not use the IDE integration at all. By requiring pip and venv knowledge we exclude exactly those inexperienced users who would most benefit from the IDE integration.

Secondly: Knowing `pip` is not that useful for using our tool. We don't even recommend our customers use `pip` to install things in their test-pack. We always recommend they install anything they can through `apt` as with that they're much more likely to end up with a working, consistent and reproducible combination of package versions.

We don't even currently mention venv in our documentation - so we can't really expect our users to even know that they don't know it.

> Home-grown wrappers that "manage" a well-known mature tool for you, always offer a bad experience.

No-one's forcing anyone to use `./stbt_rig.py setup`.  If you already know what you're doing you don't have to use it. It just gets people started on the right foot. Even if you know what you're doing actually managing a stbt venv is a pain. To do it right you need to look up on https://packages.ubuntu.com to match the versions of each package with the stbt environment. And you have to do it transitively. If you get it wrong, or don't know to do it in the first place you have to troubleshoot package versions mismatches like `astroid` and `pylint` versions diverging.

Also: We avoid the worst excesses of "home-grown wrappers" as:

1. We install nothing outside the test-pack - so we're not messing with any global state on the user's machine. If it gets in a really bad state you can delete your test pack and re-clone - it's low risk.
2. It still uses venv and pip under the hood.  It doesn't really manage this at all - it just sets it up. There's no hidden state to get out of sync if the user then goes on to install additional packages in the venv.

> But I know you're set on this so we can give it a go. You'll need to update the documentation at https://stb-tester.com/manual/ide-configuration

Yes, that's what I've got planned next - with a video on setting the environment up from scratch on Windows.

wmanley committed:
setup: Drop support for python_version = 2

Python 2 doesn't have venv and Python 2 isn't used very much any
more.  We still support stbt_rig itself being invoked with a python2
interpreter though.

wmanley committed:
setup: Clarify documentation

As suggested by @drothlis in review.
